alicloud/resource_alicloud_snapshot_policy.go

```diff
@@ -0,0 +1,166 @@
+	client := meta.(*connectivity.AliyunClient)
+	ecsService := EcsService{client}
+	snapshotPolicy, err := ecsService.DescribeSnapshotPolicy(d.Id())
+	if err != nil {
+		if NotFoundError(err) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("describe snapshot (%s) failed when read, %s", d.Id(), err)
+	}
+
+	d.Set("name", snapshotPolicy.AutoSnapshotPolicyName)
+	weekdays, err := convertJsonStringToList(snapshotPolicy.RepeatWeekdays)
+	if err != nil {
+		return fmt.Errorf("invalid RepeatDays(%s), %v", snapshotPolicy.RepeatWeekdays, err)
+	}
+	d.Set("repeat_weekdays", weekdays)
+	d.Set("retention_days", snapshotPolicy.RetentionDays)
+	timePoints, err := convertJsonStringToList(snapshotPolicy.TimePoints)
+	if err != nil {
+		return fmt.Errorf("invalid TimePoints(%s), %v", snapshotPolicy.TimePoints, err)
+	}
+	d.Set("time_points", timePoints)
+
+	return nil
+}
+
+func resourceAliyunSnapshotPolicyUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+
+	args := ecs.CreateModifyAutoSnapshotPolicyExRequest()
+	args.AutoSnapshotPolicyId = d.Id()
+	if d.HasChange("name") {
+		args.AutoSnapshotPolicyName = d.Get("name").(string)
+	}
+	if d.HasChange("repeat_weekdays") {
+		args.RepeatWeekdays = convertListToJsonString(d.Get("repeat_weekdays").(*schema.Set).List())
+	}
+	if d.HasChange("retention_dasy") {
+		args.RetentionDays = requests.NewInteger(d.Get("retention_days").(int))
+	}
+	if d.HasChange("time_points") {
+		args.TimePoints = convertListToJsonString(d.Get("time_points").(*schema.Set).List())
+	}
+	_, err := client.WithEcsClient(func(ecsClient *ecs.Client) (interface{}, error) {
+		return ecsClient.ModifyAutoSnapshotPolicyEx(args)
+	})
+	if err != nil {
+		return err
+	}
+
+	return resourceAliyunSnapshotPolicyRead(d, meta)
+}
+
+func resourceAliyunSnapshotPolicyDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*connectivity.AliyunClient)
+	ecsService := EcsService{client}
+
+	args := ecs.CreateDeleteAutoSnapshotPolicyRequest()
+	args.AutoSnapshotPolicyId = d.Id()
+
+	err := resource.Retry(DefaultTimeout*time.Second, func() *resource.RetryError {
+		_, err := client.WithEcsClient(func(ecsClient *ecs.Client) (interface{}, error) {
+			return ecsClient.DeleteAutoSnapshotPolicy(args)
+		})
+		if err != nil {
+			if NotFoundError(err) {
+				return nil
+			}
+			if IsExceptedErrors(err, SnapshotPolicyInvalidOperations) {
+				return resource.RetryableError(fmt.Errorf("delete snapshot policy (%s) failed, %#v", d.Id(), err))
+			}
+			return resource.NonRetryableError(fmt.Errorf("delete snapshot policy (%s) failed, %#v", d.Id(), err))
+		}
+
+		_, err = ecsService.DescribeSnapshotById(d.Id())
+		if err != nil {
+			if NotFoundError(err) {
+				return nil
+			}
+			return resource.NonRetryableError(fmt.Errorf("describe snapshot policy by Id failed when delete the snapshot policy (%s), %#v", d.Id(), err))
+		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
```

Comment:
Please re-build the delete method by referring the snapshot resource delete method.

Reply:
Line 140 should be `return  resource.RetryableError(err)`; line 142 should be `return return resource.NonRetryableError(err)`; line 147 should be `return WrapErrorf(err, DefaultErrorMsg, d.Id(), AlibabaCloudSdkGoERROR)`